README.md
[![](https://travis-ci.org/sakrist/VBPieChart.svg?branch=master)](https://travis-ci.org/sakrist/VBPieChart)
[![](https://img.shields.io/cocoapods/v/VBPieChart.svg?style=flat)](https://cocoapods.org/pods/VBPieChart)
[![License](http://img.shields.io/:license-mit-blue.svg)](http://doge.mit-license.org)

# VBPieChart

Animated Pie Chart control for iOS apps, based on CALayer. Very easy in use and have custom labeling.

<img src="https://raw.githubusercontent.com/sakrist/VBPieChart/master/Screenshot.png" width="50%">

## Usage


#### Create simple pi chart with VBPieChart:

```objc
VBPieChart *chart = [[VBPieChart alloc] initWithFrame:CGRectMake(10, 50, 300, 300)];
[self.view addSubview:chart];

// Setup some options:
[chart setEnableStrokeColor:YES]; 
[chart setHoleRadiusPrecent:0.3]; /* hole inside of chart */

// Prepare your data
NSArray *chartValues = @[
 @{@"name":@"Apples", @"value":@50, @"color":[UIColor redColor]},
 @{@"name":@"Pears", @"value":@20, @"color":[UIColor blueColor]},
 @{@"name":@"Oranges", @"value":@40, @"color":[UIColor orangeColor]},
 @{@"name":@"Bananas", @"value":@70, @"color":[UIColor purpleColor]}
];

// Present pie chart with animation
[chart setChartValues:chartValues animation:YES duration:0.4 options:VBPieChartAnimationFan];
```

`chartValues` needs to be defined as an array of dictionaries.<br>
Dictionary **required** to contain value for piece with key `value`.<br>
Optional:

- `name`
- `color`
- `labelColor`
- `accent`
- `strokeColor`



#### Chart from JSON

```objc
VBPieChart *chart = [[VBPieChart alloc] initWithFrame:CGRectMake(10, 50, 300, 300)];
chart.startAngle = M_PI+M_PI_2;
chart.holeRadiusPrecent = 0.5;
[self.view addSubview:chart];
NSString *json_example = @"[ {\"name\":\"first\", \"value\":\"50\", \"color\":\"#84C69B\", \"strokeColor\":\"#fff\"}, \
{\"name\":\"second\", \"value\":\"60\", \"color\":\"#FECEA8\", \"strokeColor\":\"#fff\"}, \
{\"name\":\"second\", \"value\":\"75\", \"color\":\"#F7EEBB\", \"strokeColor\":\"#fff\"}, \
{\"name\":\"second\", \"value\":\"90\", \"color\":\"#D7C1E0\", \"strokeColor\":\"#fff\"} ]";

NSData *data = [json_example dataUsingEncoding:NSUTF8StringEncoding];
NSArray *chartValues = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];

[chart setChartValues:chartValues animation:YES duration:0.5 options:VBPieChartAnimationFan];
```

#### Chart with custom label position
```objc
VBPieChart *chart = [[VBPieChart alloc] initWithFrame:CGRectMake(10, 50, 300, 300)];
chart.startAngle = M_PI+M_PI_2;
chart.holeRadiusPrecent = 0.5;
[self.view addSubview:chart];
[chart setLabelsPosition:VBLabelsPositionCustom];
[chart setLabelBlock:^CGPoint( CALayer *layer, NSInteger index) {
    CGPoint p = CGPointMake(sin(-index/10.0*M_PI)*50+50, index*30);
    return p;
}];

[chart setChartValues:@[
                         @{@"name":@"37%", @"value":@65, @"color":@"#5677fcaa", @"labelColor":@"#000"},
                         @{@"name":@"13%", @"value":@23, @"color":@"#2baf2baa", @"labelColor":@"#000"},
                         @{@"name":@"19.3%", @"value":@34, @"color":@"#b0bec5aa", @"labelColor":@"#000"},
                         @{@"name":@"30.7%", @"value":@54, @"color":@"#f57c00aa", @"labelColor":@"#000"}
                         ]
             animation:YES];
```

## Basic Documentation

`VBPieChart` is subclass of `UIView`.

#### Properties

`length`<br />
Length of circle pie. Min values is 0 and max value 2*M_PI.

`startAngle`<br />
Start angle of pie. (M_PI will make start at left side)

`holeRadiusPrecent`<br />
hole radius in % of whole radius. Values 0..1. (acual hole radius will be calculated **radius***`holeRadiusPrecent`)

`radiusPrecent` <br />
Defines the **radius**,  **full radius** = frame.size.width/2, actual **radius** = **full radius***`radiusPrecent`. Value 0..1.

`labelBlock`<br />
Block will help to redefine positions for labels.

#### Methods

Get all changed chart values back.<br />
`- (NSArray *) chartValues;`<br />

Simple setup data.<br />
`- (void) setChartValues:(NSArray *)chartValues;`<br />

Setup data to pie chart with animation or not, animation options and duration.<br />
`- (void) setChartValues:(NSArray *)chartValues animation:(BOOL)animation duration:(float)duration options:(VBPieChartAnimationOptions)options;`

Animation options:

* `VBPieChartAnimationFanAll`
* `VBPieChartAnimationGrowth`
* `VBPieChartAnimationGrowthAll`
* `VBPieChartAnimationGrowthBack`
* `VBPieChartAnimationGrowthBackAll`
* `VBPieChartAnimationFan`
* `VBPieChartAnimationTimingEaseInOut`
* `VBPieChartAnimationTimingEaseIn`
* `VBPieChartAnimationTimingEaseOut`
* `VBPieChartAnimationTimingLinear`

Change value for elemet at index. Value will be changed with animation. <br />
`- (void) setValue:(NSNumber*)value pieceAtIndex:(NSInteger)index;` <br />

Insert new piece at index. Animated.
`- (void) insertChartValue:(NSDictionary*)chartValue atIndex:(NSInteger)index;`

Remove piece at index.<br />
`- (void) removePieceAtIndex:(NSInteger)index;`<br />

Supported with ad:

<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
<!-- Github -->
<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-1267260663063670"
     data-ad-slot="2880275683"
     data-ad-format="auto"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script>


CHANGELOG.md

#### [1.2.1] - 2017-31-08
* migrate sample to latest swift 3.x

#### [1.2.0] - 2016-20-01
* fixed `removePieceAtIndex:`
* fixed `insertChartValue:atIndex:`
* `setValues:` - deleted, use `setValue:pieceAtIndex:`
* improved documentation for simple start
* added couple more examples
* JSON example
* fixed animations
* UIColor category included back for JSON reason



#### [1.1.1] - 2016-20-01
* fixed initWithFrame
* improved documentation for simple start

#### [1.1.0] - 2016-20-01
* `VBPiePiece` - made fully private, and all members are `readonly`
* from float -> double
* some little more commnets

#### [1.0.0] - 2016-19-01
* UIColor+HexColor - moved out of pod
* new documentation for `VBPieChart` class
* fixed issues with passing wrong data
* new `VBPieChartAnimationDefault`
* new `insertChartValue:atIndex:`
* reverted `- (NSArray *) chartValues;`
* fixed accent vector
* some more Unit Tests

#### [0.4.1] - 2016-18-01
* fixed crash for empty `chartValues`
* fixed `removePieceAtIndex:` if animations disabled
* implemented some simple Tests
* travis updated

#### [0.4.0] - 2016-17-01
* `chartValues` moved to private, added `setChartValues:` instead
* `setValue:forIndex:` renamed to `setValue:pieceAtIndex:`
* new `removePieceAtIndex:`
* fixed Accent with different `startAngle`
* `VBLabelBlock` now have one more parameter
* `strokeColor` and `enableStrokeColor` **DEPRECATED**
* `enableInteractive` **DEPRECATED**
* `accentVector` in VBPiePiece


Charts/VBCViewController.m
//
//  VBCViewController.m
//  Charts
//
//  Created by Volodymyr Boichentsov on 15/02/2014.
//  Copyright (c) 2014 SAKrisT. All rights reserved.
//

#import "VBCViewController.h"
#import "UIColor+HexColor.h"
#import "VBPieChart.h"



@interface VBCViewController ()

@property (nonatomic, retain) VBPieChart *chart;
@end

@implementation VBCViewController
- (IBAction)valueChangedSlider:(UISlider*)sender {
    
    float value = [sender value];
    
    switch (sender.tag) {
        case 0: {
//            [_chart setFrame:CGRectMake(100, 100, value, value)];
            NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithDictionary:_chart.chartValues];
            
            NSString *key = [[dict allKeys] objectAtIndex:sender.tag];
            [dict setObject:[NSNumber numberWithFloat:sender.value] forKey:key];
            _chart.chartValues = dict;
        }
            break;
            
        case 1:
            [_chart setRadiusPrecent:value];
            break;
        case 2:
            [_chart setHoleRadiusPrecent:value];
            break;
    }
    
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self.view setNeedsLayout];
    
    if (!_chart) {
        _chart = [[VBPieChart alloc] init];
        [self.view addSubview:_chart];
    }
    [_chart setFrame:CGRectMake(10, 50, 300, 300)];
    [_chart setEnableStrokeColor:YES];
    _chart.length = M_PI;
    _chart.startAngle = M_PI;
    
    [_chart.layer setShadowOffset:CGSizeMake(2, 2)];
    [_chart.layer setShadowRadius:3];
    [_chart.layer setShadowColor:[UIColor blackColor].CGColor];
    [_chart.layer setShadowOpacity:0.7];
    
//    [_chart setEnableInteractive:YES];
    [_chart setHoleRadiusPrecent:0.3];
    NSDictionary *chartValues = @{

                           @"firsts":  @{@"value":@35 },
                           @"first 2":  @{@"value":@20 },
                           @"first 3":  @{@"value":@10 },
                           @"first ":  @{@"value":@40},
                           @"second": @20,
                           @"third": @40,
                           @"fourth": @10,
                           };

    [_chart setChartValues:chartValues animation:YES];
//    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationGrowth];
    
	// Do any additional setup after loading the view, typically from a nib.
}


- (IBAction)setanother:(id)sender {
    NSDictionary *chartValues = @{
                           @"first": @{@"value":@50, @"color":[UIColor colorWithRed:0.55 green:0.77 blue:0.53 alpha:1.0]},
                           @"second": @{@"value":@20, @"color":[UIColor colorWithRed:0.33 green:0.17 blue:0.48 alpha:1.0]},
                           @"third": @{@"value":@40, @"color":[UIColor colorWithRed:0.81 green:0.61 blue:0.02 alpha:1.0]},
                           @"fourth": @{@"value":@10, @"color":[UIColor colorWithRed:0.31 green:0.74 blue:0.91 alpha:1.0]},
                           @"fourth 2": @{@"value":@70, @"color":[UIColor colorWithRed:0.43 green:0.02 blue:0.46 alpha:1.0]},
                           @"fourth 3": @{@"value":@65, @"color":[UIColor colorWithRed:0.00 green:0.51 blue:0.08 alpha:1.0]},
                           @"fourth 4": @{@"value":@23, @"color":[UIColor colorWithRed:0.84 green:0.66 blue:0.81 alpha:1.0]},
                           @"fourth 5": @{@"value":@34, @"color":[UIColor colorWithRed:0.73 green:0.20 blue:0.63 alpha:1.0]},
                           @"fourth 6": @{@"value":@54, @"color":[UIColor colorWithRed:0.83 green:0.20 blue:0.15 alpha:1.0]}
                           };
    
    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationGrowthAll | VBPieChartAnimationTimingEaseInOut];
}

- (IBAction)setanother1:(id)sender {
    NSDictionary *chartValues = @{
                                  @"first": @{@"value":@50, @"color":[UIColor colorWithRed:0.55 green:0.77 blue:0.53 alpha:1.0]},
                                  @"second": @{@"value":@20, @"color":[UIColor colorWithRed:0.33 green:0.17 blue:0.48 alpha:1.0]},
                                  @"third": @{@"value":@40, @"color":[UIColor colorWithRed:0.81 green:0.61 blue:0.02 alpha:1.0]},
                                  @"fourth": @{@"value":@10, @"color":[UIColor colorWithRed:0.31 green:0.74 blue:0.91 alpha:1.0]},
                                  @"fourth 2": @{@"value":@70, @"color":[UIColor colorWithRed:0.43 green:0.02 blue:0.46 alpha:1.0]},
                                  @"fourth 3": @{@"value":@65, @"color":[UIColor colorWithRed:0.00 green:0.51 blue:0.08 alpha:1.0]},
                                  @"fourth 4": @{@"value":@23, @"color":[UIColor colorWithRed:0.84 green:0.66 blue:0.81 alpha:1.0]},
                                  @"fourth 5": @{@"value":@34, @"color":[UIColor colorWithRed:0.73 green:0.20 blue:0.63 alpha:1.0]},
                                  @"fourth 6": @{@"value":@54, @"color":[UIColor colorWithRed:0.83 green:0.20 blue:0.15 alpha:1.0]}
                                  };
    
    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationGrowth];
}

- (IBAction)setanother3:(id)sender {
    NSDictionary *chartValues = @{
                                  @"first": @{@"value":@50, @"color":[UIColor colorWithRed:0.55 green:0.77 blue:0.53 alpha:1.0]},
                                  @"second": @{@"value":@20, @"color":[UIColor colorWithRed:0.33 green:0.17 blue:0.48 alpha:1.0]},
                                  @"third": @{@"value":@40, @"color":[UIColor colorWithRed:0.81 green:0.61 blue:0.02 alpha:1.0]},
                                  @"fourth": @{@"value":@10, @"color":[UIColor colorWithRed:0.31 green:0.74 blue:0.91 alpha:1.0]},
                                  @"fourth 2": @{@"value":@70, @"color":[UIColor colorWithRed:0.43 green:0.02 blue:0.46 alpha:1.0]},
                                  @"fourth 3": @{@"value":@65, @"color":[UIColor colorWithRed:0.00 green:0.51 blue:0.08 alpha:1.0]},
                                  @"fourth 4": @{@"value":@23, @"color":[UIColor colorWithRed:0.84 green:0.66 blue:0.81 alpha:1.0]},
                                  @"fourth 5": @{@"value":@34, @"color":[UIColor colorWithRed:0.73 green:0.20 blue:0.63 alpha:1.0]},
                                  @"fourth 6": @{@"value":@54, @"color":[UIColor colorWithRed:0.83 green:0.20 blue:0.15 alpha:1.0]}
                                  };
    
    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationGrowthBackAll | VBPieChartAnimationTimingEaseInOut];
}

- (IBAction)setanother4:(id)sender {
    NSDictionary *chartValues = @{
                                  @"first": @{@"value":@50, @"color":[UIColor colorWithRed:0.55 green:0.77 blue:0.53 alpha:1.0]},
                                  @"second": @{@"value":@20, @"color":[UIColor colorWithRed:0.33 green:0.17 blue:0.48 alpha:1.0]},
                                  @"third": @{@"value":@40, @"color":[UIColor colorWithRed:0.81 green:0.61 blue:0.02 alpha:1.0]},
                                  @"fourth": @{@"value":@10, @"color":[UIColor colorWithRed:0.31 green:0.74 blue:0.91 alpha:1.0]},
                                  @"fourth 2": @{@"value":@70, @"color":[UIColor colorWithRed:0.43 green:0.02 blue:0.46 alpha:1.0]},
                                  @"fourth 3": @{@"value":@65, @"color":[UIColor colorWithRed:0.00 green:0.51 blue:0.08 alpha:1.0]},
                                  @"fourth 4": @{@"value":@23, @"color":[UIColor colorWithRed:0.84 green:0.66 blue:0.81 alpha:1.0]},
                                  @"fourth 5": @{@"value":@34, @"color":[UIColor colorWithRed:0.73 green:0.20 blue:0.63 alpha:1.0]},
                                  @"fourth 6": @{@"value":@54, @"color":[UIColor colorWithRed:0.83 green:0.20 blue:0.15 alpha:1.0]}
                                  };
    
    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationGrowthBack];
}


- (IBAction)setanother2:(id)sender {
    NSDictionary *chartValues = @{
                                  @"first": @{@"value":@50, @"color":[UIColor colorWithRed:0.55 green:0.77 blue:0.53 alpha:1.0]},
                                  @"second": @{@"value":@20, @"color":[UIColor colorWithRed:0.33 green:0.17 blue:0.48 alpha:1.0]},
                                  @"third": @{@"value":@40, @"color":[UIColor colorWithRed:0.81 green:0.61 blue:0.02 alpha:1.0]},
                                  @"fourth": @{@"value":@10, @"color":[UIColor colorWithRed:0.31 green:0.74 blue:0.91 alpha:1.0]},
                                  @"fourth 2": @{@"value":@70, @"color":[UIColor colorWithRed:0.43 green:0.02 blue:0.46 alpha:1.0]},
                                  @"fourth 3": @{@"value":@65, @"color":[UIColor colorWithRed:0.00 green:0.51 blue:0.08 alpha:1.0]},
                                  @"fourth 4": @{@"value":@23, @"color":[UIColor colorWithRed:0.84 green:0.66 blue:0.81 alpha:1.0]},
                                  @"fourth 5": @{@"value":@34, @"color":[UIColor colorWithRed:0.73 green:0.20 blue:0.63 alpha:1.0]},
                                  @"fourth 6": @{@"value":@54, @"color":[UIColor colorWithRed:0.83 green:0.20 blue:0.15 alpha:1.0]}
                                  };
    
    [_chart setChartValues:chartValues animation:YES options:VBPieChartAnimationFan];
}


- (IBAction) changeLenght:(id)sender {
    if (_chart.length < M_PI*2-0.01) {
        _chart.length = M_PI*2;
        _chart.startAngle = 0;
    } else {
        _chart.length = M_PI;
        _chart.startAngle = M_PI;
    }
    
    [_chart setChartValues:_chart.chartValues animation:YES];
    
}


//- (void) touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
//
//    UITouch *t = [[touches allObjects] lastObject];
//    CGPoint p1 = [t previousLocationInView:self.view];
//    CGPoint p2 = [t locationInView:self.view];
//    
//    CGPoint delta;
//    delta.x = p1.x - p2.x;
//    delta.y = p1.y - p2.y;
//    
//    _chart.layer.transform = CATransform3DRotate(_chart.layer.transform, delta.y * M_PI / 180.0f, 1, 0, 0);
////    self.layer.transform = CATransform3DRotate(self.layer2.transform, delta.y * M_PI / 180.0f, 1, 0, 0);
//
//    _chart.layer.transform = CATransform3DRotate(_chart.layer.transform, delta.x * M_PI / 180.0f, 0, -1, 0);
////    self.layer2.transform = CATransform3DRotate(self.layer2.transform, delta.x * M_PI / 180.0f, 0, -1, 0);
//
//    
//}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end
